Blog post Error Handling Tradeoffs and Crashing in Production (#272)
coveo/source.coveo.com: _posts/2023-04-28-error-handling-tradeoffs-and-crashing-in-production.md

This is one of the reasons some businesses like Google famously decided not to use them.
Actually, when you throw an exception, the so-called stack unwinding happens which ends up running a lot of code (for instance the `destructor`s).
Finally, exceptions can be a little bit too transparent.
It is easy to forget to handle an exception that we _had to_ handle for our program to work properly.

## Throw an exception or crash?

A question that I asked myself for a long time is: should I `assert` or should I `throw`.
Yes, `assert` should be for conditions that are _always true_, but my experience as a developer taught me humility: what I think is always true is usually true... until it becomes false!
Sometime, this happened in production code, where the `assert`s were removed.
In this case, the bug can propagate until something bad happen, usually quite far from the place where the `assert` would have failed if it was a debug build.
This behavior made release build hard to debug: perhaps replacing `assert` with `throw` was the answer?

> I gave it a lot of thought, I read a lot, listen to a lot of talks, and I came to a surprising conclusion: sometime, you should crash the program in production.

Author: Thanks! I do this error sometimes...

![image](https://user-images.githubusercontent.com/8481892/232126915-fabb73a0-59bc-43fc-a019-2335af4a825e.png)

Of course, something should restart the program.
That being said, there's still the risk of getting stuck in a [crashloop](https://stackoverflow.com/a/52215388/3068259).
So, in which condition is it worth risking a crashloop in production?
When your program is in a corrupted state.

The behavior of a corrupted program is no longer predictable.
It is clear with this definition that a corrupted program is dangerous, and the rise of security incidents made the danger of such programs very visible.
For instance, consider the C++ code

> ```c

Reviewer: cpp ?

Author: Nope. C'est du bon vieux C!

Reviewer: ouin mais ton texte dit : consider the C++ code

assert(i >= 0); // always true
assert(i < arraySize); // always true
array[i] += userInput;
```

If, for some reason that we don't understand, the variable `i` becomes an arbitrary value, this code could, depending on how it is compiled, result in an arbitrary code execution from the user.
Hence, we should definitely not remove those checks in production code.
This is with this principle in mind that the [C++ contract proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0380r1.pdf) was written.
If the predicate of the contract is `false`, the program will crash, both in release build and in debug build.

Why not throw an exception instead of crashing?
For two reasons.
First, we assume that nobody down the stack will have enough information to handle the exception: we are in a situation that we thought as impossible and for all we know, the stack itself could be corrupted!

> The second reason is that, as mentioned before, throwing an exception actually execute a lot of code, which is not safe in a context where the program is corrupted.

Reviewer: Is this an actual issue? my understanding was that stack unwinding wasn't as dangerous, but it's really far in my memory 😓

Reviewer: Oh this is in the context of the contract, nvm

Author: > Oh this is in the context of the contract, nvm
Not really. See [this paper](https://www.ndss-symposium.org/wp-content/uploads/2023/02/ndss2023_s295_paper.pdf) for instance. I'll add a reference to it in the text.

## Checked exceptions

> To make it less likely that developers forget to deal with exceptions that they have to handle, Java introduced checked exception.

Reviewer 2: ```suggestion
To make it less likely that developers forget to deal with exceptions that they have to handle, Java introduced checked exceptions.
```

If a method throws a checked exception, for instance an `IOException`, it _must_ declare it in its signature.

> When a method `DoX()` doesn't declare that it's throwing an exception, the compiler enforces it.

Reviewer 2: detail but java is lowerCamelCase so
```suggestion
When a method `doX()` doesn't declare that it's throwing an exception, the compiler enforces it.
```

> Hence, if `DoX()` calls `DoIO() throws IOException`, it needs to do it in a `try` block and to `catch` the `IOException` that could be thrown.

Reviewer 2: ```suggestion
Hence, if `doX()` calls `doIO() throws IOException`, it needs to do it in a `try` block and to `catch` the `IOException` that could be thrown. Alternatively, it could declare it in its own signature to pass the checked exception to the parent caller.
```

The powerful compile-time check described in the previous paragraph comes with tradeoffs.
Since checked exceptions are part of the API of the methods, it means that a change in the exceptions thrown will result in an API break.
This is, of course, required in a method where we want, for instance, to catch every exception, like in the `DoX()` example above.
However, there are some context where methods don't need to know if an exception is thrown and where we would like transparent propagation, like in the function `getBoldHtml` mentioned before.
As explained above, this is actually one of the strengths of the exception concept: it allows separation of concerns between business logic and error handling.

> Java programmers invented workarounds to allow recovering this transparent propagation with checked exceptions, like using the generic `throws Exception`, but some of these workarounds were equivalent to stop using checked exceptions altogether.

Reviewer: I guess we could also talk about RuntimeException here for completeness sakes and maybe mention that this is how Spring decided to use that strategy.

Is it possible to have the best of both worlds, robust compile time checks and transparent propagation?

# The return of the return codes

After the advent of checked exceptions, some programming languages and some libraries started to return, for each function and method call that might fail, an object that contains the result of the computation as well as information about the error that might have occurred.

> Rust uses this approach, and it uses it in a way that allows developers to benefit from the advantage of checked exception _and_ of transparent propagation.

Reviewer 3: ```suggestion
Rust uses this approach, and uses it in a way that allows developers to benefit from the advantage of checked exceptions _and_ of transparent propagation.
```

> Rust uses this approach, and it uses it in a way that allows developers to benefit from the advantage of checked exception _and_ of transparent propagation.

Reviewer: Rust and Go

Reviewer: (but it wouldn't add that much value to the text tbh)

Reviewer 3: Midly related to this, but I feel this paragraph could use a small clarification.
>some programming languages and some libraries started to return, for each function and method call that might fail, an object that contains the result of the computation as well as information about the error that might have occurred.

Reading this in isolation, this sounds a bit like Go without mentioning it explicitly (i.e. fallible functions return both a value _and_ optionally, an error) and the following part:

>Rust uses this approach, and it uses it in a way that allows developers to benefit from the advantage of checked exception _and_ of transparent propagation.

 doesn't really highlight how Rust's approach is different from using C/Go with a linter: that the object returned by fallible functions contains _either_ the return value _or_ an error, and the caller _must_ take in to account the failure modes in order to get to the underlying return value.

A few sentences down, you do mention that rust forces the user to handle the error, but I think it might be nice to highlight how exactly it does it 😄

Author: @dalloriam I reworked the paragraph. Tell me what you think about the reworked version. :slightly_smiling_face:

Author: (Je vais attendre ton feedback avant de merger)

In Rust, every error must be handled explicitly, _but_, the operator `?` allows propagating the error to the calling function.
For instance, consider the function[^2]

```rust
fn read_username_from_file() -> Result<String, io::Error> {
    let mut username_file = File::open("hello.txt")?;
    let mut username = String::new();
    username_file.read_to_string(&mut username)?;
    Ok(username)
}
```

> In this example, the `String` that contains the `username` will usually be returned, but if the call `File::open("hello.txt")` or `read_to_string(&mut username)` fail and return an `io::Error` object, the `?` operator will return this object right away.

Reviewer 3: ```suggestion
In this example, the `String` that contains the `username` will usually be returned, but if the call `File::open("hello.txt")` or `read_to_string(&mut username)` fails and returns an `io::Error` object, the `?` operator will return this object right away.
```

An example of main function that calls `read_user_from_file` is

```rust
fn main() {
    let user_result = read_username_from_file();

    match user_result {
        Ok(username) => println!("Hello, {username}!");,
        Err(error) => panic!("Problem opening the file: {:?}", error),
    };
}
```

The more astute readers might notice that the transparent propagation problem is not solved yet.
Indeed, `read_username_from_file` needs to know that `File::open` and `read_to_string` can return `io::Error`, and then it needs to declare it in the return type in its signature `-> Result<String, io::Error>`.
Hence, if `File::open` started to throw a new error type, for instance `io::ErrorInterface`, the signature of `read_username_from_file` would need to be modified.
Luckily for us, there are libraries that handle the error abstraction problem, like [thiserror](https://docs.rs/thiserror/latest/thiserror/) and [miette](https://docs.rs/miette/latest/miette/).
These libraries were implemented in part thanks to the powerful Rust macro system.

# Closing thoughts

From C to Rust, the facilities that we can use to handle errors have evolved quite a bit.

> What used to look like mutual exclusive tradeoffs are now available to us at the same time, making it easier to design programs that are as reliable as exciting!

Reviewer: Est-ce que ça vaudrait la peine de faire un retour sur ton _it depends_ plus haut ? e.g. 
>and the actual best solution to your problem is "it depends" and what actually matter is understanding what context you are operating in

Author: Hmmm. Je sais pas. Les languages avaient tous des tradeoffs dans la manière où ils géraient les exceptions, donc oui, "it used to depend" parce qu'avant on devait choisir les tradeoff les plus importants pour nous, mais maintenant, on a l'opportunité d'avoir des languages qui apprenent des erreurs des vieux languages et qui ont mis des mécanismes en place qui évitent la plupart des pièges...

Author: Fait que mainenant, it depends du language que tu veux utiliser I guess. :laughing:
